### PR TITLE
Added ts-prune to detect unused code

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
[ts-prune](https://github.com/nadeesha/ts-prune/) is a nifty tool that detects unused code.

~Unfortunately it has a bug in the latest versions (https://github.com/nadeesha/ts-prune/issues/151) that make it not suitable for use in CI just yet, but at least the older versions mostly don't have bugs.~ ✅

### CODE REVIEW

File: packages/worker/src/utils.ts [hunk 1]
```diff
@@ -1,8 +1,5 @@
-export const sleep = (ms: number): Promise<void> => new Promise((resolve) => setTimeout(resolve, ms));
-export const identity = <T>(x: T): T => x;
```
Comment (JoshuaKGoldberg) [inline packages/worker/src/utils.ts:2]: `identity` was unused. `sleep` was only used in a test file so I moved it there.

File: packages/worker/src/utils.ts [hunk 2]
```diff
@@ -1,8 +1,5 @@
...
-
-export const GiB = 1024 ** 3;
 export const MiB = 1024 ** 2;
+export const GiB = MiB * 1024;
```
Comment (JoshuaKGoldberg) [inline packages/worker/src/utils.ts:2]: `MiB` was unused but factors into `GiB` nicely.

File: tsconfig.prune.json [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+  "files": [
+    "./packages/activity/src/index.ts",
+    "./packages/client/src/index.ts",
+    "./packages/common/src/index.ts",
+    "./packages/create-project/src/index.ts",
+    "./packages/meta/src/index.ts",
+    "./packages/worker/src/index.ts",
+    "./packages/worker/src/workflow/workflow-worker-thread.ts",
+    "./packages/workflow/src/index.ts"
+  ]
```
Comment (JoshuaKGoldberg) [inline tsconfig.prune.json:17]: These `files` act as the entry points for unused code detection. In other words, code things that are transitively imported by these files are counted as used.

File: packages/worker/src/rxutils.ts [hunk 4]
```diff
@@ -8,19 +8,6 @@ interface StateAndOptionalOutput<T, O> {
 
 export type StateAndOutput<T, O> = Required<StateAndOptionalOutput<T, O>>;
 
-export function mapWithState<T, I, O>(
```
Comment (JoshuaKGoldberg) [inline packages/worker/src/rxutils.ts:11]: Totally unused 🤘

File: packages/worker/src/index.ts [hunk 5]
```diff
@@ -11,7 +11,7 @@
 export { State, Worker, ActivityOptions, DataConverter, defaultDataConverter, RetryOptions, errors } from './worker';
 export { WorkerOptions, CompiledWorkerOptions } from './worker-options';
 export { ServerOptions, TLSConfig } from './server-options';
-export { Core } from './core';
+export { Core, CompiledCoreOptions, CoreOptions, TelemetryOptions } from './core';
```
Comment (JoshuaKGoldberg) [inline packages/worker/src/index.ts:14]: These options interfaces are used in `Core`'s functions but not exported. Perhaps not required for unused code detection, but a nice to export.

Comment (bergundy) [reply packages/worker/src/index.ts:14]: Agree

File: packages/worker/src/rxutils.ts (hunk 4 above)
Comment (bergundy) [reply packages/worker/src/rxutils.ts:11]: yup

**Review by bergundy — COMMENTED** [review]
Overall lgtm, thanks!

Let's add this as a CI step.

**Comment by JoshuaKGoldberg** [comment]
Moving to https://github.com/temporalio/sdk-typescript/pull/405 👍

**Pull request closed without merging**